### PR TITLE
Update DS simulation modes

### PR DESCRIPTION
**Pull request by daria305**
DS simulation can now works in two modes:
 - one for accidental double spend: `SimulationMode= "Accidental"`
 - one for intentional double spend: `SimulationMode= "Adversary"`

### CODE REVIEW

File: simulation/parser.go [hunk 1]
```diff
@@ -143,24 +158,38 @@ func parseAdversaryConfig(adversaryDelays *string, adversaryTypes *string, adver
...
-	// make sure mana and delays are only defined when adversary type is provided and have the same length
+	// make sure mana, nodeCounts and delays are only defined when adversary type is provided and have the same length
 	if len(config.AdversaryDelays) != 0 && len(config.AdversaryDelays) != len(config.AdversaryTypes) {
 		config.AdversaryDelays = []int{}
 	}
 	if len(config.AdversaryMana) != 0 && len(config.AdversaryMana) != len(config.AdversaryTypes) {
 		config.AdversaryMana = []float64{}
 	}
+	if len(config.AdversaryNodeCounts) != 0 && len(config.AdversaryNodeCounts) != len(config.AdversaryTypes) {
+		config.AdversaryNodeCounts = []int{}
```
Comment (bingyanglin) [inline simulation/parser.go:181]: Might be good to add some Info-level logs to inform the user the setting might be wrong, or the user might get unexpected results due to wrong settings.

File: network/groups.go [hunk 2]
```diff
@@ -83,72 +84,68 @@ func NewAdversaryGroups() (groups AdversaryGroups) {
...
-			}
-			groupIndexQueue.Offer(groupIndex)
-		}
-	}
-	for groupIndex, group := range *g {
-		log.Infof("GroupId: %d, AdversaryType: %d, Node count: %d, q: %.5f, adNodeList: %v", groupIndex, group.AdversaryType, len(group.NodeIDs),
-			group.GroupMana/totalWeight, group.NodeIDs)
+func (g *AdversaryGroups) updateGroupMana() {
+	for _, group := range *g {
+		group.GroupMana = group.TargetManaPercentage * float64(config.NodesTotalWeight)
```
Comment (bingyanglin) [inline network/groups.go:148]:
```suggestion
		group.GroupMana = group.TargetManaPercentage * float64(config.NodesTotalWeight) / 100.0
```
Do we need to divide it by 100.0 here?

File: simulation/parser.go [hunk 3]
```diff
@@ -143,24 +158,38 @@ func parseAdversaryConfig(adversaryDelays *string, adversaryTypes *string, adver
...
-	// maximum 3 adversary types allowed for one simulation, one per color
-	if len(config.AdversaryTypes) > 3 {
+	// no adversary if colors are not provided
+	if len(config.AdversaryInitColors) != len(config.AdversaryTypes) {
 		config.AdversaryTypes = []int{}
 	}
 
-	// make sure mana and delays are only defined when adversary type is provided and have the same length
+	// make sure mana, nodeCounts and delays are only defined when adversary type is provided and have the same length
 	if len(config.AdversaryDelays) != 0 && len(config.AdversaryDelays) != len(config.AdversaryTypes) {
```
Comment (bingyanglin) [inline simulation/parser.go:174]:
```suggestion
	if len(config.AdversaryDelays) != 0 && len(config.AdversaryDelays) != len(config.AdversaryTypes) {
		log.Warnf("The AdversaryDelays count is not equal to AdversaryTypes count!")	
```
We can use warning log to warn the user the setting might be wrong.

File: simulation/parser.go [hunk 4]
```diff
@@ -143,24 +158,38 @@ func parseAdversaryConfig(adversaryDelays *string, adversaryTypes *string, adver
...
 		config.AdversaryDelays = []int{}
 	}
 	if len(config.AdversaryMana) != 0 && len(config.AdversaryMana) != len(config.AdversaryTypes) {
```
Comment (bingyanglin) [inline simulation/parser.go:177]:
```suggestion
	if len(config.AdversaryMana) != 0 && len(config.AdversaryMana) != len(config.AdversaryTypes) {
		log.Warnf("The AdversaryMana count is not equal to the AdversaryTypes count!")
```

File: simulation/parser.go [hunk 5]
```diff
@@ -143,24 +158,38 @@ func parseAdversaryConfig(adversaryDelays *string, adversaryTypes *string, adver
...
 		config.AdversaryMana = []float64{}
 	}
+	if len(config.AdversaryNodeCounts) != 0 && len(config.AdversaryNodeCounts) != len(config.AdversaryTypes) {
```
Comment (bingyanglin) [inline simulation/parser.go:180]:
```suggestion
	if len(config.AdversaryNodeCounts) != 0 && len(config.AdversaryNodeCounts) != len(config.AdversaryTypes) {
		log.Warnf("The AdversaryNodeCounts count is not equal to the AdversaryTypes count!")	
```

**Review by bingyanglin — CHANGES_REQUESTED** [review]
- Need to investigate the adversary weight to check whether their percentage are correctly set.
- Might be good to add some Info-level logs to inform the user the setting might be wrong, or the user might get unexpected results due to wrong settings. (Though they can check that in the configuration dump file)

File: network/groups.go [hunk 6]
```diff
@@ -83,72 +84,68 @@ func NewAdversaryGroups() (groups AdversaryGroups) {
 			delay = config.AdversaryDelays[i]
 		}
 
-		groups = append(groups, &AdversaryGroup{
-			NodeIDs:       make([]int, 0),
-			TargetMana:    targetMana,
-			Delay:         time.Millisecond * time.Duration(delay),
-			AdversaryType: ToAdversaryType(configAdvType),
-		})
+		if len(config.AdversaryNodeCounts) > 0 {
```
Comment (piotrm50) [inline network/groups.go:87]: what happens when number of adversary node counts and adversary types is different? shoudn't it panic?

**Review by bingyanglin — APPROVED** [review]
LGTM

File: network/groups.go [hunk 7]
```diff
@@ -173,34 +170,29 @@ func randomWeightIndex(weights []uint64, count int) (randomWeights []int) {
...
-				minLocation = i
-			}
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region Accidental ///////////////////////////////////////////////////////////////////////////////////////////////////
+
+func GetAccidentalIssuers(network *Network) []*Peer {
+	peers := make([]*Peer, 0)
+	randomCount := 0
+	for i := 0; i < len(config.AccidentalMana); i++ {
```
Comment (piotrm50) [inline network/groups.go:180]: Don't we also want to select issuers from a given index, sorted by the Weight? That means we could get a node from the middle of the distribution and use that as well.

File: network/groups.go (hunk 6 above)
Comment (daria305) [reply network/groups.go:87]: I have it covered during parsing this way:
```
		if len(config.AdversaryNodeCounts) != 0 && len(config.AdversaryNodeCounts) != len(config.AdversaryTypes) {
		config.AdversaryNodeCounts = []int{}
	}
```

File: network/groups.go (hunk 7 above)
Comment (daria305) [reply network/groups.go:180]: why not, might be usefull in the future, I'll add that

File: network/groups.go (hunk 6 above)
Comment (piotrm50) [reply network/groups.go:88]: So if the length is not the same for all the parameters, it falls back to using 2 random color issuers?

**Review by piotrm50 — APPROVED** [review]

File: network/groups.go (hunk 6 above)
Comment (daria305) [reply network/groups.go:88]: in this case the default values (node count = 1) for AdversaryNodeCounts will be used with warning

**Pull request merged**